### ISSUE TITLE
After the successful construction in wsl2, an error is reported by running the. / kodev run instruction

### ISSUE THREAD

**Issue by aaaaaaaaala** [koreader/koreader#6947]
[*] Current time: 12/04/20-14:38:41
 [*] Version: v2020.11
12/04/20-14:38:41 INFO  platform thirdparty apps
-> Goldendict (goldendict), role: dict, available: false

Segmentation fault
*** Segmentation fault
Register dump:

 RAX: 0000000000000000   RBX: 00007fff13938f94   RCX: 0000000000000000
 RDX: 0000000000000001   RSI: 0000000000000000   RDI: 00007fd81b2f1968
 RBP: 00007fd81b2a2150   R8 : 0000000000000000   R9 : 0000000000000000
 R10: 0000000000000032   R11: 000056297172c9b8   R12: 00007fff13938f8c
 R13: 00007fff13938f94   R14: 00007fff13938f8c   R15: 00007fd81b2a2150
 RSP: 00007fff13938ed8

**Comment by Frenzie** [comment]
Please try to mostly follow the issue template. For example, WSL2 is only one part of the story. It's WSL2 *plus* a particular version of Ubuntu/Debian/Fedora/etc.

I'd regard WSL2 as unsupported. Nice if it happens to run, PRs to fix any issues welcome.

I'm not sure if lack of an X server could cause a segmentation fault?

**Comment by aaaaaaaaala** [comment]
> Please try to mostly follow the issue template. For example, WSL2 is only one part of the story. It's WSL2 _plus_ a particular version of Ubuntu/Debian/Fedora/etc.
> 
> I'd regard WSL2 as unsupported. Nice if it happens to run, PRs to fix any issues welcome.
> 
> I'm not sure if lack of an X server could cause a segmentation fault?

thank you! It's Ubuntu 18.04. Xeyes works normally, but there's a problem with koreader.
![image](https://user-images.githubusercontent.com/75469690/101135008-2b7aa280-3646-11eb-8003-2ccc757dd655.png)

**Comment by aaaaaaaaala** [comment]
> > Please try to mostly follow the issue template. For example, WSL2 is only one part of the story. It's WSL2 _plus_ a particular version of Ubuntu/Debian/Fedora/etc.
> > I'd regard WSL2 as unsupported. Nice if it happens to run, PRs to fix any issues welcome.
> > I'm not sure if lack of an X server could cause a segmentation fault?
> 
> thank you! It's Ubuntu 18.04. Xeyes works normally, but there's a problem with koreader.
> ![image](https://user-images.githubusercontent.com/75469690/101135008-2b7aa280-3646-11eb-8003-2ccc757dd655.png)

Is wsl1 supported?

**Comment by Frenzie** [comment]
WSL1 has worse compatibility with Linux than WSL2, so let's say it's supported by however much less that compatibility is. ;-)

KOReader does run on both WSL1 and WSL2, see the instructions on https://github.com/koreader/virdevenv#windows-workflow for one possible method.

What I mean in so many words is that on average, any problem you run into is most likely to be a Windows problem, and with that you're mostly if not entirely on on your own. If you know how to work gdb you can try to figure out which library is segfaulting. Actually I'm surprised `catchsegv` isn't giving you more output.

What method are you using to run an X server?

**Comment by pazos** [comment]
> What method are you using to run an X server?

This.

Probably SDL does not find its way to X. The last line before segfaulting is part of device early init. SDL init logs are expected just after that

**Comment by aaaaaaaaala** [comment]
> WSL1 has worse compatibility with Linux than WSL2, so let's say it's supported by however much less that compatibility is. ;-)
> 
> KOReader does run on both WSL1 and WSL2, see the instructions on https://github.com/koreader/virdevenv#windows-workflow for one possible method.
> 
> What I mean in so many words is that on average, any problem you run into is most likely to be a Windows problem, and with that you're mostly if not entirely on on your own. If you know how to work gdb you can try to figure out which library is segfaulting. Actually I'm surprised `catchsegv` isn't giving you more output.
> 
> What method are you using to run an X server?

Run xrdp and then use windows remote desktop connection。

**Comment by aaaaaaaaala** [comment]
> > WSL1 has worse compatibility with Linux than WSL2, so let's say it's supported by however much less that compatibility is. ;-)
> > KOReader does run on both WSL1 and WSL2, see the instructions on https://github.com/koreader/virdevenv#windows-workflow for one possible method.
> > What I mean in so many words is that on average, any problem you run into is most likely to be a Windows problem, and with that you're mostly if not entirely on on your own. If you know how to work gdb you can try to figure out which library is segfaulting. Actually I'm surprised `catchsegv` isn't giving you more output.
> > What method are you using to run an X server?
> 
> Run xrdp and then use windows remote desktop connection。

Are these things printed?

`Segmentation fault
*** Segmentation fault
Register dump:

 RAX: 0000000000000000   RBX: 00007ffc2b81d714   RCX: 0000000000000000
 RDX: 0000000000000001   RSI: 0000000000000000   RDI: 00007fbf5cb54968
 RBP: 00007fbf5cac1178   R8 : 0000000000000000   R9 : 0000000000000000
 R10: 0000000000000032   R11: 00005566bc8579b8   R12: 00007ffc2b81d70c
 R13: 00007ffc2b81d714   R14: 00007ffc2b81d70c   R15: 00007fbf5cac1178
 RSP: 00007ffc2b81d658

 RIP: 00007fbf5b5152cf   EFLAGS: 00010212

 CS: 0033   FS: 0000   GS: 0000

 Trap: 0000000e   Error: 00000004   OldMask: 00000000   CR2: 00000008

 FPUCW: 0000037f   FPUSW: 00000000   TAG: 00000000
 RIP: 00000000   RDP: 00000000

 ST(0) 0000 0000000000000000   ST(1) 0000 0000000000000000
 ST(2) 0000 0000000000000000   ST(3) 0000 0000000000000000
 ST(4) 0000 0000000000000000   ST(5) 0000 0000000000000000
 ST(6) 0000 0000000000000000   ST(7) 0000 0000000000000000
 mxcsr: 1fa1
 XMM0:  00000000000000000000000000000000 XMM1:  00000000000000000000000000000000
 XMM2:  00000000000000000000000000000000 XMM3:  00000000000000000000000000000000
 XMM4:  00000000000000000000000000000000 XMM5:  00000000000000000000000000000000
 XMM6:  00000000000000000000000000000000 XMM7:  00000000000000000000000000000000
 XMM8:  00000000000000000000000000000000 XMM9:  00000000000000000000000000000000
 XMM10: 00000000000000000000000000000000 XMM11: 00000000000000000000000000000000
 XMM12: 00000000000000000000000000000000 XMM13: 00000000000000000000000000000000
 XMM14: 00000000000000000000000000000000 XMM15: 00000000000000000000000000000000

Backtrace:
/usr/lib/x86_64-linux-gnu/libSDL2.so(+0xe12cf)[0x7fbf5b5152cf]
/usr/lib/x86_64-linux-gnu/libSDL2.so(+0x428c2)[0x7fbf5b4768c2]
./luajit(+0x25082)[0x5566ba995082]
./luajit(+0x44c37)[0x5566ba9b4c37]
./luajit(+0x2094c)[0x5566ba99094c]
./luajit(+0x22c55)[0x5566ba992c55]
./luajit(lua_call+0x1f)[0x5566ba9822af]
./luajit(+0x5095d)[0x5566ba9c095d]
./luajit(+0x22c55)[0x5566ba992c55]
./luajit(lua_call+0x1f)[0x5566ba9822af]
./luajit(+0x5095d)[0x5566ba9c095d]
./luajit(+0x22c55)[0x5566ba992c55]
./luajit(lua_pcall+0x5e)[0x5566ba982312]
./luajit(+0x7c1e)[0x5566ba977c1e]
./luajit(+0x84d6)[0x5566ba9784d6]
./luajit(+0x8bec)[0x5566ba978bec]
./luajit(+0x22c55)[0x5566ba992c55]
./luajit(lua_cpcall+0x23)[0x5566ba982365]
./luajit(main+0x42)[0x5566ba978c6d]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7fbf5bb9bbf7]
./luajit(_start+0x2a)[0x5566ba9778ca]

Memory map:

5566ba970000-5566ba9ea000 r-xp 00000000 08:10 393773 /home/mo/tem/koreader/base/build/x86_64-linux-gnu-debug/luajit
5566babea000-5566babec000 r--p 0007a000 08:10 393773 /home/mo/tem/koreader/base/build/x86_64-linux-gnu-debug/luajit
5566babec000-5566babed000 rw-p 0007c000 08:10 393773 /home/mo/tem/koreader/base/build/x86_64-linux-gnu-debug/luajit
`

**Comment by aaaaaaaaala** [comment]
> > What method are you using to run an X server?
> 
> This.
> 
> Probably SDL does not find its way to X. The last line before segfaulting is part of device early init. SDL init logs are expected just after that

thank you! So, how can I solve this problem?

**Comment by pazos** [comment]
>  So, how can I solve this problem?

A quick search gives me https://stackoverflow.com/questions/58187388/using-sdl-with-windows-subsystem-for-linux

You need to be able to do your own research. If not please use Virtualbox.

**Comment by aaaaaaaaala** [comment]
> > So, how can I solve this problem?
> 
> A quick search gives me https://stackoverflow.com/questions/58187388/using-sdl-with-windows-subsystem-for-linux
> 
> You need to be able to do your own research. If not please use Virtualbox.

thank you!

**Comment by Frenzie** [comment]
I gave it a quick try with VcXsrv with the most important bits as described in https://medium.com/javarevisited/using-wsl-2-with-x-server-linux-on-windows-a372263533c3

There's a minor screen painting issue (i.e., the window's only black or white), but otherwise it works fine. A minor hiccup not mentioned in that guide is that you apparently need to allow the server full access on *public* networks to communicate with WSL.

In summary, there seems to be no issue other than that running graphical apps in WSL is slightly tricky.

**Comment by aaaaaaaaala** [comment]
> I gave it a quick try with VcXsrv with the most important bits as described in https://medium.com/javarevisited/using-wsl-2-with-x-server-linux-on-windows-a372263533c3
> 
> There's a minor screen painting issue (i.e., the window's only black or white), but otherwise it works fine. A minor hiccup not mentioned in that guide is that you apparently need to allow the server full access on _public_ networks to communicate with WSL.
> 
> In summary, there seems to be no issue other than that running graphical apps in WSL is slightly tricky.

thank you!
